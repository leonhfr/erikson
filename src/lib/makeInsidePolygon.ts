// Packages.
import * as _ from 'lodash';
import * as turfHelpers from '@turf/helpers';
import union from '@turf/union';

// Internal.
import { Geometry } from './Geometry';

// Code.
export const makeInsidePolygon = (
  boxes: Array<Array<turfHelpers.BBox>>
):
  | turfHelpers.Feature<turfHelpers.Polygon | turfHelpers.MultiPolygon>
  | undefined => {
  const polygons: Array<
    turfHelpers.Feature<turfHelpers.Polygon | turfHelpers.MultiPolygon>
  > = [];

  for (const colBboxes of boxes) {
    const mergedBoxes = mergeBoxes(colBboxes);

    if (mergedBoxes.length === 1) {
      polygons.push(Geometry.bboxToPolygon(mergedBoxes[0]));
      continue;
    }

    polygons.push(Geometry.bboxesToMultiPolygon(mergedBoxes));
  }

  return _.reduce(polygons, (acc, val) => union(acc, val));
};

export const mergeBoxes = (boxes: Array<turfHelpers.BBox>) => {
  let accumulator: Array<turfHelpers.BBox> = [];
  return _.reduce(boxes, mergeIteratee, accumulator);
};

export const mergeIteratee = (
  boxes: Array<turfHelpers.BBox>,
  value: turfHelpers.BBox
): Array<turfHelpers.BBox> => {
  if (!boxes.length) {
    return [value];
  }

  const [minX, minY, maxX, maxY] = value;
  const lastBbox = boxes[boxes.length - 1];

  if (minY === lastBbox[3]) {
    return [
      ...boxes.slice(0, -1),
      [minX, lastBbox[1], maxX, maxY] as turfHelpers.BBox,
    ];
  }

  return [...boxes, value];
};
