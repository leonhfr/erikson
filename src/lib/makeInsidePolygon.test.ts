// Packages.
import * as turfHelpers from '@turf/helpers';

// Internal.
import { makeInsidePolygon, mergeBoxes } from './makeInsidePolygon';

// Code.
describe('makeInsidePolygon', () => {
  it('should correctly compute the polygon', () => {
    const boxes: Array<Array<turfHelpers.BBox>> = [
      [
        [0, 0, 1, 1],
        [0, 1, 1, 2],
        [0, 2, 1, 3],
        [0, 4, 1, 5],
        [0, 6, 1, 7],
        [0, 7, 1, 8],
      ],
      [
        [1, 1, 2, 2],
        [1, 2, 2, 3],
        [1, 3, 2, 4],
        [1, 4, 2, 5],
        [1, 5, 2, 6],
        [1, 6, 2, 7],
      ],
      [[2, 2, 3, 3], [2, 4, 3, 6], [2, 6, 3, 7]],
    ];
    const result = makeInsidePolygon(boxes);
    expect(result).toMatchSnapshot();
  });
});

describe('mergeBoxes', () => {
  it('should correctly reduce the boxes', () => {
    const boxes: Array<turfHelpers.BBox> = [
      [0, 0, 1, 1],
      [0, 1, 1, 2],
      [0, 2, 1, 3],
      [0, 4, 1, 5],
      [0, 6, 1, 7],
      [0, 7, 1, 8],
    ];
    const result = mergeBoxes(boxes);
    expect(result).toMatchSnapshot();
  });
});
