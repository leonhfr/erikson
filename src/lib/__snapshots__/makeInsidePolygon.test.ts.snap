// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeInsidePolygon should correctly compute the polygon 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          0,
          0,
        ],
        Array [
          1,
          0,
        ],
        Array [
          1,
          1,
        ],
        Array [
          2,
          1,
        ],
        Array [
          2,
          2,
        ],
        Array [
          3,
          2,
        ],
        Array [
          3,
          3,
        ],
        Array [
          2,
          3,
        ],
        Array [
          2,
          4,
        ],
        Array [
          3,
          4,
        ],
        Array [
          3,
          7,
        ],
        Array [
          2,
          7,
        ],
        Array [
          1,
          7,
        ],
        Array [
          1,
          8,
        ],
        Array [
          0,
          8,
        ],
        Array [
          0,
          6,
        ],
        Array [
          1,
          6,
        ],
        Array [
          1,
          5,
        ],
        Array [
          0,
          5,
        ],
        Array [
          0,
          4,
        ],
        Array [
          1,
          4,
        ],
        Array [
          1,
          3,
        ],
        Array [
          0,
          3,
        ],
        Array [
          0,
          0,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;

exports[`mergeBoxes should correctly reduce the boxes 1`] = `
Array [
  Array [
    0,
    0,
    1,
    3,
  ],
  Array [
    0,
    4,
    1,
    5,
  ],
  Array [
    0,
    6,
    1,
    8,
  ],
]
`;
